### DISCUSSION TITLE
Using Vertical Slice Architecture

### DISCUSSION THREAD

**Discussion by mehdihadeli** [kgrzybek/modular-monolith-with-ddd#225] [Ideas] [answered]
Hi @kgrzybek,
In your project, you used a vertical slice architecture in each module separately, correct?
But for each `slice`, we have all onion physical layers also, and our slices spread between this physical layers. Is it ok (actually mix of vertical and onion architecture)? In the jimmy boggard [sample](https://github.com/jbogard/ContosoUniversityDotNetCore-Pages), all related things for each slice exists in a feature folder without using an n-tire architecture.

**Comment by AndreiGanichev** [comment, marked as answer]
Hi, @mehdihadeli!

As far as I know _Vertical slice architecture_ doesn’t have a strong definition. It’s about an **idea of grouping code according to business functionality**. The idea can be applied on different levels of abstraction. Slice can be a module, a group of use cases under the same Aggregate and actually a use case by itself. I recommend you to read this quite interesting [thread](https://twitter.com/kamgrzybek/status/1161634025712340992?s=20):

> I see vertical slices in hierarchical form. Module is the root (Orders), which contains other smaller slices (OrderPlacement, OrderManagement) which contains smaller slices to the level of the one atomic operation like PlaceOrder, CancelOrder etc.. [tweet by Kamil Grzybek](https://twitter.com/kamgrzybek/status/1161634025712340992?s=20)

We can consider [Feature Folders](http://www.kamilgrzybek.com/design/feature-folders/) as vertical slices inside each module.

Slices don’t actually contradict layers and as you can see in the project can successfully live together. **Slices** make our architecture screaming and easy to understand. **Layers** provide abilities for encapsulation(with ```internal``` keyword) and allow us to manage dependencies.

In [this talk](https://www.youtube.com/watch?v=SUiWfhAhgQw&t=1s) Jimmy Bogard said that one of the reasons to use vertical slice architecture had been over complicated repositories. It was because these repositories served both read and write requests. CQRS solves this problem by dividing Read and Write models. So you can see how simple and readable repositories in the project are.

Vertical slice architecture can be a starting point and can be evolved later when an application become more sophisticated:
> we can start simple (Transaction Script) and simply refactor to the patterns that emerges from code smells we see in the business logic. [Vertical slice architecture by Jimmy Bogard](https://jimmybogard.com/vertical-slice-architecture/).

> I would introduce modules when/if my application grows beyond a single BC. [tweet by Jimmy Bogard](https://twitter.com/jbogard/status/1161641870205890561?s=20)

We can find close ideas(introduce layers only where there is complex business logic of different levels) in Modular Monolith: Domain-Centric Design article of Kamil Grzybek:
> ...within the same module there may be both more and less complicated functionalities. In this case, we should also respect each functionality separately. ...the application of layers should not be a global decision for a given module – each use case should be considered separately.

And finally I would suggest to read [Layers, hexagons, features and components](http://www.codingthearchitecture.com/2016/04/25/layers_hexagons_features_and_components.html) article of Simon Braun where he looks at different architecture styles in terms of encapsulation.

**Reply by mehdihadeli** [reply]
@AndreiGanichev Thanks for the explanation
